### PR TITLE
[SPARK-40508][SQL] Treat unknown partitioning as UnknownPartitioning

### PR DESCRIPTION
**Pull request by tedyu**
### What changes were proposed in this pull request?
When running spark application against spark 3.3, I see the following :
```
java.lang.IllegalArgumentException: Unsupported data source V2 partitioning type: CustomPartitioning
    at org.apache.spark.sql.execution.datasources.v2.V2ScanPartitioning$$anonfun$apply$1.applyOrElse(V2ScanPartitioning.scala:46)
    at org.apache.spark.sql.execution.datasources.v2.V2ScanPartitioning$$anonfun$apply$1.applyOrElse(V2ScanPartitioning.scala:34)
    at org.apache.spark.sql.catalyst.trees.TreeNode.$anonfun$transformDownWithPruning$1(TreeNode.scala:584)
```
The `CustomPartitioning` works fine with Spark 3.2.1
This PR proposes to relax the code and treat all unknown partitioning the same way as that for `UnknownPartitioning`.

### Why are the changes needed?
3.3.0 doesn't seem to warrant such behavioral change (from that of 3.2.1 release).

### Does this PR introduce _any_ user-facing change?
This would allow user's custom partitioning to continue to work with 3.3.x releases.

### How was this patch tested?
Existing test suite.

### CODE REVIEW

**Comment by tedyu** [comment]
cc @sunchao

**Comment by sunchao** [comment]
@tedyu could you share some background information? if `CustomPartitioning` is handled the same way as `UnknownPartitioning`, why can't you just use the latter instead?

**Comment by tedyu** [comment]
If I subclass `UnknownPartitioning` directly, I would get this compilation error:
```
[error] /nfusr/dev-server/zyu/spark-cassandra-connector/connector/src/main/scala/com/datastax/spark/connector/datasource/CassandraScanBuilder.scala:327:92: not enough arguments for constructor UnknownPartitioning: (x$1: Int)org.apache.spark.sql.connector.read.partitioning.UnknownPartitioning.
[error] Unspecified value parameter x$1.
[error] case class CassandraPartitioning(partitionKeys: Array[String], numPartitions: Int) extends UnknownPartitioning {
[error]                                                                                            ^
[error] one error found
```

**Comment by sunchao** [comment]
Can you directly report `UnknownPartitioning` to Spark?

**Comment by tedyu** [comment]
If custom partitioning reports `UnknownPartitioning` to Spark and can keep 3.2.1 behavior, that means the current check is not desired.

**Comment by tedyu** [comment]
I have run the test using Cassandra Spark connector and modified Spark (with this patch).

The test passes (without modification to Cassandra Spark connector or client code).

**Comment by sunchao** [comment]
I guess this PR make sense. @tedyu could you:
- create a Spark JIRA for this issue? and update the PR title to reflect it?
- add a warning message too? clients may expect Spark to use the partitioning they reported and could be surprised that Spark internally ignores it, so a warning message would be helpful for them to debug.

I think the best solution is for connectors such as Cassandra to adopt the new API, otherwise they could see severe performance penalties.

**Comment by tedyu** [comment]
@sunchao 
Please take another look.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2ScanPartitioningAndOrdering.scala [hunk 1]
```diff
@@ -54,8 +55,9 @@ object V2ScanPartitioningAndOrdering extends Rule[LogicalPlan] with SQLConfHelpe
             }
           }
         case _: UnknownPartitioning => None
-        case p => throw new IllegalArgumentException("Unsupported data source V2 partitioning " +
-            "type: " + p.getClass.getSimpleName)
+        case p =>
+          logWarning("Spark ignores the partitioning. Please use KeyGroupedPartitioning for better performance")
```
Comment (sunchao) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2ScanPartitioningAndOrdering.scala:59]: nit: could we also log the class name here? e.g.,
```
Spark ignores partitioning ${p.getClass.getSimpleName}. Please use KeyGroupedPartitioning for better performance
```

**Review by sunchao — APPROVED** [review]

**Comment by tedyu** [comment]
@sunchao 
https://github.com/tedyu/spark/runs/8459534296 shows that all tests have passed.

**Comment by tedyu** [comment]
@sunchao 
Can this PR be merged ?

**Comment by sunchao** [comment]
Thanks! merged to master. @tedyu could you create a PR for branch-3.3 as well?

**Review by dongjoon-hyun — COMMENTED** [review]
Thank you, @tedyu and @sunchao .
If this was a regression from 3.2.1, could you provide a unit test case please before backporting to branch-3.3?

Also, cc @wangyum since he is the release manager for Apache Spark 3.3.1.

**Comment by sunchao** [comment]
Yea good point. A unit test would be nice.

**Pull request closed without merging**